src/web/services/search.service.ts

```diff
@@ -169,47 +157,44 @@ export class SearchService {
   }
 
   getPrivileges(
-    coursesWithSections: SearchStudentsTable[],
+    coursesWithStudents: SearchStudentsListRowTable[],
   ): Observable<InstructorPrivilege[]> {
-    if (coursesWithSections.length === 0) {
+    if (coursesWithStudents.length === 0) {
       return of([]);
     }
-    return forkJoin(
-      coursesWithSections.map((course: SearchStudentsTable) => {
-        return course.sections.map((section: StudentListSectionData) => {
-          return this.instructorService.loadInstructorPrivilege({
-            courseId: course.courseId,
-            sectionName: section.sectionName,
-          });
-        });
-      }).reduce(
-        (acc: Observable<InstructorPrivilege>[], val: Observable<InstructorPrivilege>[]) =>
-        acc.concat(val),
-        [],
-      ),
-    );
+    const privileges: any[] = [];
+    for (const course of coursesWithStudents) {
+      const distinctSections: string[] = Array.from(
+        new Set(course.students.map((studentModel: StudentListRowModel) => studentModel.student.sectionName)));
+      const instructorPrivileges: Observable<InstructorPrivilege>[] = distinctSections.map((section: string) =>
+        this.instructorService.loadInstructorPrivilege({ courseId: course.courseId, sectionName: section }));
+      for (const studentModel of course.students) {
+        privileges.push(instructorPrivileges[distinctSections.indexOf(studentModel.student.sectionName)]);
+      }
```

Comment:
You can probably use a `Record<string, Observable<InstructorPrivilege>>` to make things faster. Map from section name to the observable.